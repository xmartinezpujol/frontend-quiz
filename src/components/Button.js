import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import COLOR_PALETTE from '../Constants';

const shape = {
  round: '200px',
  square: 0,
  default: 6,
};

const size = {
  small: {
    fontSize: 14,
  },
  medium: {
    fontSize: 18,
  },
  large: {
    fontSize: 22,
  },
};

const template = (props) => {
  const templateRemake = {
    default: {
      background: COLOR_PALETTE[props.type],
      border: 0,
      color: COLOR_PALETTE.nori,
    },
    dynamic: {
      color: COLOR_PALETTE.purewhite,
      position: 'relative',
      zIndex: 1,
      border: 0,
      background: `linear-gradient(to right, ${props.gradient[0]}, ${props.gradient[1]})`,
      '&:after': {
        content: "''",
        bottom: '-10px',
        left: '5%',
        height: '100%',
        width: '90%',
        position: 'absolute',
        background: 'inherit',
        top: '0.5rem',
        filter: 'blur(0.4rem)',
        opacity: 0.7,
        zIndex: -1,
      },
    },
    outlined: {
      backgroundColor: COLOR_PALETTE.rice,
      border: '2px solid',
      borderColor: COLOR_PALETTE[props.bordercolor],
      color: props.bordercolor
        ? COLOR_PALETTE[props.bordercolor]
        : COLOR_PALETTE.night,
      ':hover:enabled': {
        background: (props.bordercolor === COLOR_PALETTE.nori || !props.bordercolor)
          ? COLOR_PALETTE.night
          : COLOR_PALETTE.rice,
        color: (props.bordercolor === COLOR_PALETTE.nori || !props.bordercolor)
          ? COLOR_PALETTE.rice
          : props.bordercolor,
      },
    },
    link: {
      backgroundColor: COLOR_PALETTE[props.type],
      border: 0,
      color: props.bordercolor
        ? COLOR_PALETTE[props.bordercolor]
        : COLOR_PALETTE.night,
      ':hover:enabled': {
        boxShadow: 'none',
        color: props.bordercolor
          ? props.bordercolor === COLOR_PALETTE.nori
            ? COLOR_PALETTE.rice
            : props.bordercolor
          : COLOR_PALETTE.nori,
      },
    },
  };

  const theme = templateRemake[props.template ? props.template : 'default'];
  const templateMod = Object.assign(
    {},
    theme,
  );

  return templateMod;
};

const Icon = styled.i(
  {
    display: 'inline-flex',
    padding: '18px 3px',
  },
  props => ({
    fontSize: props.iconSize,
    width: props.iconWidth,
    fontWeight: props.iconWeight,
  }),
);

const ButtonBox = styled.button(
  {
    fontFamily: 'Poppins, sans-serif',
    margin: '0 10px',
    cursor: 'pointer',
    display: 'inline-flex',
    alignItems: 'center',
    padding: '10px 35px',
    transition: '0.25s cubic-bezier(0.17, 0.67, 0.52, 0.97)',
    WebkitTapHighlightColor: 'rgba(255, 255, 255, 0)',
    ':hover:enabled': {
      boxShadow: '0 7px 14px rgba(50,50,93,.1), 0 3px 6px rgba(0,0,0,.08)',
    },
    ':focus': {
      outline: 0,
    },
    ':disabled': {
      opacity: 0.5,
      cursor: 'not-allowed',
    },
  },
  props => ({
    ...template(props),
    ...size[props.size],
    borderRadius: shape[props.shape],
    fontWeight: props.bold ? 700 : 500,
    minWidth: props.fullWidth ? '100%' : 'initial',
    textTransform: props.uppercase ? 'uppercase' : 'initial',
    transform: props.noanim ? 'none' : 'translateY(-1px)',
    ':active:enabled': {
      transform: props.noanim ? 'none' : 'translateY(1px)',
    },
    ':hover': {
      transform: props.noanim ? 'none' : 'translateY(-5px)',
    },
    ':hover:after': {
      bottom: props.noanim ? 'inherit' : '-10px',
    },
  }),
);

const Button = props => (
  <ButtonBox {...props}>
    {props.icon !== ''
    && (
      <Icon
        iconWeight={props.iconWeight}
        iconSize={props.iconSize}
        iconWidth={props.iconWidth}
        className={`${props.iFont} ${props.iFont}-${props.icon}`}
        aria-hidden="true"
      />
    )}
    {props.children}
  </ButtonBox>
);

Button.propTypes = {
  type: PropTypes.string,
  template: PropTypes.string,
  color: PropTypes.string,
  bordercolor: PropTypes.string,
  bold: PropTypes.bool,
  shape: PropTypes.string,
  size: PropTypes.string,
  icon: PropTypes.string,
  iconSize: PropTypes.number,
  iconWeight: PropTypes.number,
  iconWidth: PropTypes.number,
  iFont: PropTypes.string,
  fullWidth: PropTypes.bool,
  noanim: PropTypes.bool,
  uppercase: PropTypes.bool,
  gradient: PropTypes.arrayOf(PropTypes.string),
};

Button.defaultProps = {
  type: 'wasabi',
  template: 'default',
  color: COLOR_PALETTE.nori,
  bordercolor: COLOR_PALETTE.night,
  bold: false,
  shape: 'default',
  size: 'medium',
  icon: '',
  iconSize: 22,
  iconWeight: 300,
  iconWidth: 22,
  iFont: 'fa',
  fullWidth: false,
  noanim: false,
  uppercase: false,
  gradient: ['#0fd850', '#35FF69'],
};

export default Button;
