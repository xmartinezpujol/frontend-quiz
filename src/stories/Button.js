import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean, select, number } from '@storybook/addon-knobs';

import Button from '../components/Button';
import View from '../components/View';

const optionsTemplate = {
  default: 'default',
  outlined: 'outlined',
  link: 'link',
  dynamic: 'dynamic',
};
const optionsBorder = {
  night: 'night',
  consumer: 'consumer',
  business: 'business',
  positive: 'positive',
  temporary: 'temporary',
  negative: 'negative',
};

storiesOf('Button', module)
  .addDecorator(withKnobs)
  .add('default', (() => {
    const defaultTeamplate = 'default';
    const defaultborderColor = 'black';

    const label = 'Theme';
    const label2 = 'Border Color';

    const template = select(label, optionsTemplate, defaultTeamplate);
    const borderColor = select(label2, optionsBorder, defaultborderColor);

    return (
      <View
        style={{ width: '100vw', height: '100vh' }}
        container
        align="center"
        justify="center"
        direction="column"
      >
        <Button
          border={boolean('Border', true)}
          color={borderColor}
          bordercolor={borderColor}
          bold={boolean('Bold', true)}
          disabled={boolean('Disabled', false)}
          uppercase={boolean('Uppercase', true)}
          template={template}
          noanim={boolean('No Animation', false)}
        >
          {text('text 1', 'Button')}
        </Button>
      </View>
    );
  }))
  .add('with icon', () => {
    const defaultTeamplate = 'default';
    const defaultborderColor = 'black';

    const label = 'Theme';
    const label2 = 'Border Color';

    const template = select(label, optionsTemplate, defaultTeamplate);
    const borderColor = select(label2, optionsBorder, defaultborderColor);

    return (
      <View
        style={{ width: '100vw', height: '100vh' }}
        container
        align="center"
        justify="center"
        direction="column"
      >
        <Button
          iFont={text('Icon Font', 'fa')}
          icon={text('Icon name', 'bomb')}
          iconSize={number('Icon fontSize', 22)}
          iconWidth={number('Icon width (for paddings)', 40)}
          border={boolean('Border', true)}
          color={borderColor}
          bordercolor={borderColor}
          bold={boolean('Bold', true)}
          disabled={boolean('Disabled', false)}
          uppercase={boolean('Uppercase', true)}
          template={template}
          noanim={boolean('No Animation', false)}
        >
          {text('text 1', 'Button')}
        </Button>
      </View>
    );
  })
  .add('with icon only', () => {
    const defaultTeamplate = 'default';
    const defaultborderColor = 'black';

    const label = 'Theme';
    const label2 = 'Border Color';

    const template = select(label, optionsTemplate, defaultTeamplate);
    const borderColor = select(label2, optionsBorder, defaultborderColor);

    return (
      <View
        style={{ width: '100vw', height: '100vh' }}
        container
        align="center"
        justify="center"
        direction="column"
      >
        <Button
          iFont={text('Icon Font', 'fa')}
          icon={text('Icon name (FA)', 'comment')}
          iconSize={number('Icon fontSize', 20)}
          style={{ padding: '0px 17px' }}
          border={boolean('Border', true)}
          shape="round"
          color={borderColor}
          bordercolor={borderColor}
          bold={boolean('Bold', true)}
          disabled={boolean('Disabled', false)}
          uppercase={boolean('Uppercase', true)}
          template={template}
          noanim={boolean('No Animation', false)}
        />
      </View>
    );
  });
