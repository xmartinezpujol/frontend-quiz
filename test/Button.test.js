import React from 'react';
import { shallow, configure } from 'enzyme';
import serializer, { toHaveStyleRule } from 'jest-glamor-react';
import Adapter from 'enzyme-adapter-react-16';

import Button from '../src/components/Button';

expect.addSnapshotSerializer(serializer);
expect.extend({ toHaveStyleRule });

configure({ adapter: new Adapter() });

describe('Button', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Button theme="default">
        Button
      </Button>,
    );
  });
  it('renders correctly', () => {
    expect(wrapper.render()).toHaveStyleRule('min-width', 'initial');
    expect(wrapper.render()).toMatchSnapshot();
  });
  it('fullWidth Button should be at min-width 100%', () => {
    wrapper.setProps({ fullWidth: true });
    expect(wrapper.render()).toHaveStyleRule('min-width', '100%');
  });
  it('has no theme, should be default', () => {
    wrapper.setProps({ theme: false });
    expect(wrapper.render()).toHaveStyleRule('border', '0');
    expect(wrapper.render()).toHaveStyleRule('color', 'black');
  });
  it('has icon, should render icon font classes', () => {
    wrapper.setProps({
      theme: false,
      iconWeight: 700,
      iconSize: 22,
      iconWidth: 20,
      iFont: 'fa',
      icon: 'bomb',
    });
    expect(wrapper.find('.fa').hasClass('fa-bomb'));
  });
});
