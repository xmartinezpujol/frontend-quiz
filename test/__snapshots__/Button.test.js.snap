// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "css-5kp7wz",
      "icon": "",
      "type": "yellow",
    },
    "children": Array [
      Object {
        "data": "Button",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "icon": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "icon": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
